modules/launcher/src/test/java/org/mule/runtime/module/launcher/log4j2/LoggerContextCacheTestCase.java

```diff
@@ -0,0 +1,101 @@
+  }
+
+  @After
+  public void after() {
+    loggerContextCache.dispose();
+  }
+
+  @Test
+  public void when_recursive_MuleLoggerContext_instantiation_expect_RecursiveLoggerContextInstantiationException_and_recovery() {
+    LoggerContext expectedLoggerContext = mock(MuleLoggerContext.class);
+    when(contextSelector.buildContext(currentClassLoader))
+        .thenAnswer(invocation -> loggerContextCache.getLoggerContext(currentClassLoader))
+        .thenReturn(expectedLoggerContext);
+    LoggerContext actualLoggerContext;
+    try {
+      actualLoggerContext = loggerContextCache.getLoggerContext(currentClassLoader);
+      fail("Recursive instantiation should throw RecursiveInstantiationException");
+    } catch (MuleRuntimeException e) {
+      assertThat("Exception should be caused by RecursiveContextInstantiation", e.getClass(),
+                 IsEqual.equalTo(RecursiveLoggerContextInstantiationException.class));
+      actualLoggerContext = loggerContextCache.getLoggerContext(currentClassLoader);
+    }
+    assertThat("Invalid LoggerContext", actualLoggerContext, IsEqual.equalTo(expectedLoggerContext));
+  }
+
+  @Test
+  public void when_MuleRuntimeException_during_MuleLoggerContext_instantiation_expect_recovery() {
+    LoggerContext expectedLoggerContext = mock(MuleLoggerContext.class);
+    when(contextSelector.buildContext(currentClassLoader))
+        .thenThrow(MuleRuntimeException.class)
+        .thenAnswer(invocation -> expectedLoggerContext);
+    LoggerContext actualLoggerContext = null;
+    try {
+      loggerContextCache.getLoggerContext(currentClassLoader);
+    } catch (MuleRuntimeException mre) {
+      actualLoggerContext = loggerContextCache.getLoggerContext(currentClassLoader);
+    }
+    assertThat("Invalid loggerContext", actualLoggerContext, IsEqual.equalTo(expectedLoggerContext));
+  }
+
+  @Test
+  public void when_MuleLoggerContext_instantiation_expect_cache_store() {
+    LoggerContext firstExpectedLoggerContext = mock(MuleLoggerContext.class);
+    LoggerContext secondExpectedLoggerContext = mock(MuleLoggerContext.class);
+    when(contextSelector.buildContext(currentClassLoader)).thenReturn(firstExpectedLoggerContext);
+    when(contextSelector.buildContext(regionClassLoader)).thenReturn(secondExpectedLoggerContext);
+    loggerContextCache.getLoggerContext(currentClassLoader);
+    ClassUtils.withContextClassLoader(regionClassLoader, () -> {
+      loggerContextCache.getLoggerContext(regionClassLoader);
+    });
+    assertThat("Additional or missing LoggerContext instances found in cache", loggerContextCache.getAllLoggerContexts(),
+               IsCollectionWithSize.hasSize(2));
+    assertThat("Cache should contain firstExpectedLogger", loggerContextCache.getAllLoggerContexts(),
+               IsCollectionContaining.hasItem(firstExpectedLoggerContext));
+    assertThat("Cache should contain secondExpectedLogger", loggerContextCache.getAllLoggerContexts(),
```

Comment:
it is not necessary to use the first parameter. The exception raised by hamcrest is declarative enough.